### PR TITLE
Update non-Patient data sync via composition

### LINKED ISSUE #1307: [Quest] - Load only Binary & Parameter resources before Login
**Name of feature to enhance**
Loading app-specific configurations based on Composition resource

**Description of feature**
Only Binary and Parameter Resources specified within the Composition resource should be loaded before login. These do not require authentication.

**Describe the enhancement**
Based on the discussions within issue [#1279](https://github.com/opensrp/fhircore/issues/1279) Non-Patient Data (Questionnaire, StructureMaps, QuestionnaireResponse, Binary and Parameter Resources) will be synced through the Composition Resource.
Loading of Questionnaire, StructureMaps, and QuestionnaireResponse resources however requires authentication. 
This change will ensure that only Binary and Parameter Resources are loaded before login. The other resources will be loaded/synced after login

### PR DESCRIPTION
**Pull request by roywanyaga**
**IMPORTANT: Where possible all PRs must be linked to a Github issue**

Fixes #1307  

**Checklist**
- [ ] I have written **Unit tests** for any new feature(s) and edge cases for bug fixes
- [ ] I have added any strings visible on UI components to the `strings.xml` file
- [ ] I have updated the  [CHANGELOG.md](./CHANGELOG.md) file for any notable changes to the codebase
- [ ] I have run `./gradlew spotlessApply` and `./gradlew spotlessCheck` to check my code follows the project's style guide
- [ ] I have built and run the fhircore app to verify my change fixes the issue and/or does not break the app

### CODE REVIEW

File: android/engine/src/main/assets/configs/default/config_composition.json [hunk 1]
```diff
@@ -90,6 +90,26 @@
...
+    },
+    {
+      "title": "Add Family Member",
+      "focus": {
+        "reference": "Questionnaire/82953",
+        "identifier": {
+          "value": "family-member-registration"
+        }
+      },
+      "mode": "working"
```
Comment (dubdabasoduba) [inline android/engine/src/main/assets/configs/default/config_composition.json:112]: @roywanyaga This composition should be added to the eCBIS folder here https://github.com/opensrp/fhir-resources. 

Why are the StructureMaps used by the Questionnaires missing from the composition?

File: android/engine/src/main/java/org/smartregister/fhircore/engine/configuration/ConfigurationRegistry.kt [hunk 2]
```diff
@@ -96,6 +103,11 @@ constructor(
...
       .also { if (it == null) configsLoadedCallback(false) }
       ?.section
+      ?.filter {
+        it.focus.reference.split("/")[0].let { resourceType ->
+          resourceType == ResourceType.Parameters.name || resourceType == ResourceType.Binary.name
+        }
+      }
       ?.forEach {
         // each section in composition represents workflow
         // { "title": "register configuration",
```
Comment (dubdabasoduba) [inline android/engine/src/main/java/org/smartregister/fhircore/engine/configuration/ConfigurationRegistry.kt:113]: @roywanyaga Remove dead code

File: android/engine/src/main/java/org/smartregister/fhircore/engine/configuration/ConfigurationRegistry.kt [hunk 3]
```diff
@@ -96,6 +103,11 @@ constructor(
       .searchCompositionByIdentifier(appId)
       .also { if (it == null) configsLoadedCallback(false) }
       ?.section
+      ?.filter {
+        it.focus.reference.split("/")[0].let { resourceType ->
```
Comment (dubdabasoduba) [inline android/engine/src/main/java/org/smartregister/fhircore/engine/configuration/ConfigurationRegistry.kt:107]: Could have the `"/"` on a constant? It's used multiple times on the code

Comment (dubdabasoduba) [reply android/engine/src/main/java/org/smartregister/fhircore/engine/configuration/ConfigurationRegistry.kt:107]: Also, should we check where index 0 has anything after the split?

File: android/engine/src/main/java/org/smartregister/fhircore/engine/configuration/ConfigurationRegistry.kt [hunk 4]
```diff
@@ -169,6 +186,32 @@ constructor(
...
   }
 
+  fun fetchNonWorkflowConfigResources() {
+    CoroutineScope(dispatcherProvider.io()).launch {
+      try {
+        Timber.i("Fetching non-workflow resources for app $appId")
+        repository
+          .searchCompositionByIdentifier(appId)
+          ?.section
+          ?.groupBy { it.focus.reference.split("/")[0] }
```
Comment (dubdabasoduba) [inline android/engine/src/main/java/org/smartregister/fhircore/engine/configuration/ConfigurationRegistry.kt:196]: What happens if index 0 is empty?

File: android/engine/src/main/java/org/smartregister/fhircore/engine/configuration/ConfigurationRegistry.kt [hunk 5]
```diff
@@ -169,6 +186,32 @@ constructor(
...
+          ?.entries
+          ?.filterNot {
+            it.key == ResourceType.Binary.name || it.key == ResourceType.Parameters.name
+          }
+          ?.forEach { entry: Map.Entry<String, List<Composition.SectionComponent>> ->
+            val ids = entry.value.joinToString(",") { it.focus.extractId() }
+            val rPath = entry.key + "?${Composition.SP_RES_ID}=$ids"
```
Comment (dubdabasoduba) [inline android/engine/src/main/java/org/smartregister/fhircore/engine/configuration/ConfigurationRegistry.kt:203]: Please use more descriptive variable names. This helps make the code readable now that we do not document all methods.

**Review by dubdabasoduba — CHANGES_REQUESTED** [review]

File: android/engine/src/main/java/org/smartregister/fhircore/engine/configuration/ConfigurationRegistry.kt (hunk 2 above)
Comment (roywanyaga) [reply android/engine/src/main/java/org/smartregister/fhircore/engine/configuration/ConfigurationRegistry.kt]: This has been updated. The comments have been provided as javadoc

File: android/engine/src/main/java/org/smartregister/fhircore/engine/configuration/ConfigurationRegistry.kt (hunk 3 above)
Comment (roywanyaga) [reply android/engine/src/main/java/org/smartregister/fhircore/engine/configuration/ConfigurationRegistry.kt]: This has been updated. The safe null check has been provided.

File: android/engine/src/main/java/org/smartregister/fhircore/engine/configuration/ConfigurationRegistry.kt (hunk 4 above)
Comment (roywanyaga) [reply android/engine/src/main/java/org/smartregister/fhircore/engine/configuration/ConfigurationRegistry.kt]: safe null check has been implemented

File: android/engine/src/main/java/org/smartregister/fhircore/engine/configuration/ConfigurationRegistry.kt (hunk 5 above)
Comment (roywanyaga) [reply android/engine/src/main/java/org/smartregister/fhircore/engine/configuration/ConfigurationRegistry.kt]: This has been updated

File: android/engine/src/main/assets/configs/default/config_composition.json (hunk 1 above)
Comment (roywanyaga) [reply android/engine/src/main/assets/configs/default/config_composition.json:112]: [pull request to update the composition and other debug resources for eCBIS](https://github.com/opensrp/fhir-resources/pull/65) has been created:

File: android/engine/src/main/java/org/smartregister/fhircore/engine/ui/appsetting/AppSettingViewModel.kt [hunk 6]
```diff
@@ -87,14 +87,18 @@ constructor(
...
-          entry: Map.Entry<String, List<Composition.SectionComponent>> ->
-          val ids = entry.value.joinToString(",") { it.focus.extractId() }
-          val rPath = entry.key + "?${Composition.SP_RES_ID}=$ids"
-          fhirResourceDataSource.loadData(rPath).entry.forEach {
-            defaultRepository.save(it.resource)
+        composition
+          .section
+          .groupBy { it.focus.reference.split("/")[0] }
+          .entries
+          .filter { it.key == ResourceType.Binary.name || it.key == ResourceType.Parameters.name }
```
Comment (pld) [inline android/engine/src/main/java/org/smartregister/fhircore/engine/ui/appsetting/AppSettingViewModel.kt:94]: try `in`

**Review by pld — COMMENTED** [review]
a couple comments nothing major, these tests look great!

File: android/engine/src/main/java/org/smartregister/fhircore/engine/configuration/ConfigurationRegistry.kt [hunk 7]
```diff
@@ -88,27 +95,28 @@ constructor(
...
 
+  /**
+   *
+   * each section in composition represents workflow { "title": "register configuration", "mode":
+   * "working", "focus": { "reference": "Binary/11111", "identifier: { "value": "registration" } } }
+   *
+   * A workflow point would be mapped like "workflowPoint": "registration", "resource":
+   * "RegisterViewConfiguration", "classification": "patient_register", "description": "register
+   * configuration"
+   */
```
[the remark this answers is not in the document]
Comment (roywanyaga) [reply android/engine/src/main/java/org/smartregister/fhircore/engine/configuration/ConfigurationRegistry.kt:122]: This has been effected

File: android/engine/src/main/java/org/smartregister/fhircore/engine/configuration/ConfigurationRegistry.kt [hunk 8]
```diff
@@ -88,27 +95,28 @@ constructor(
...
     this.appId = appId
 
-    // appId is identifier of Composition
     repository
       .searchCompositionByIdentifier(appId)
       .also { if (it == null) configsLoadedCallback(false) }
       ?.section
+      ?.filter {
+        it.focus.reference?.split(TYPE_REFERENCE_DELIMITER)?.get(0).let { resourceType ->
+          resourceType == ResourceType.Parameters.name || resourceType == ResourceType.Binary.name
```
[the remark this answers is not in the document]
Comment (roywanyaga) [reply android/engine/src/main/java/org/smartregister/fhircore/engine/configuration/ConfigurationRegistry.kt]: This has been effected

File: android/engine/src/main/java/org/smartregister/fhircore/engine/configuration/ConfigurationRegistry.kt [hunk 9]
```diff
@@ -136,6 +144,11 @@ constructor(
         .use { it.readText() }
         .decodeResourceFromString<Composition>()
         .section
+        .filter {
```
[the remark this answers is not in the document]
Comment (roywanyaga) [reply android/engine/src/main/java/org/smartregister/fhircore/engine/configuration/ConfigurationRegistry.kt]: This has been effected

File: android/engine/src/main/java/org/smartregister/fhircore/engine/configuration/ConfigurationRegistry.kt [hunk 10]
```diff
@@ -169,6 +182,34 @@ constructor(
       ?.also { configsLoadedCallback(true) }
   }
 
+  fun fetchNonWorkflowConfigResources() {
```
[the remark this answers is not in the document]
Comment (roywanyaga) [reply android/engine/src/main/java/org/smartregister/fhircore/engine/configuration/ConfigurationRegistry.kt:205]: This has been effected

File: android/engine/src/main/java/org/smartregister/fhircore/engine/configuration/ConfigurationRegistry.kt [hunk 11]
```diff
@@ -169,6 +182,34 @@ constructor(
...
+    CoroutineScope(dispatcherProvider.io()).launch {
+      try {
+        Timber.i("Fetching non-workflow resources for app $appId")
+        repository
+          .searchCompositionByIdentifier(appId)
+          ?.section
+          ?.groupBy { it.focus.reference?.split(TYPE_REFERENCE_DELIMITER)?.get(0) ?: "" }
+          ?.entries
+          ?.filterNot { sectionEntry ->
+            sectionEntry.key == ResourceType.Binary.name ||
```
[the remark this answers is not in the document]
Comment (roywanyaga) [reply android/engine/src/main/java/org/smartregister/fhircore/engine/configuration/ConfigurationRegistry.kt]: This has been effected

File: android/engine/src/main/java/org/smartregister/fhircore/engine/ui/appsetting/AppSettingViewModel.kt (hunk 6 above)
Comment (roywanyaga) [reply android/engine/src/main/java/org/smartregister/fhircore/engine/ui/appsetting/AppSettingViewModel.kt:94]: This has been effected

**Review by dubdabasoduba — APPROVED** [review]

**Pull request merged**